webapp/doc_parser.py

```diff
@@ -0,0 +1,234 @@
+            index_topic["post_stream"]["posts"][0]["cooked"],
+            features="lxml",
+        )
+
+        # Parse URL & redirects mappings (get warnings)
+        self.url_map, url_warnings = self._parse_url_map(raw_index_soup)
+        self.redirect_map, redirect_warnings = self._parse_redirect_map(
+            raw_index_soup
+        )
+        self.warnings = url_warnings + redirect_warnings
+
+        # Get body and navigation HTML
+        self.index_document = self.parse_topic(index_topic)
+        index_soup = BeautifulSoup(
+            self.index_document["body_html"], features="lxml"
+        )
+        self.index_document["body_html"] = str(
+            self._get_preamble(index_soup, break_on_title="Navigation")
+        )
+
+        # Parse navigation
+        self.navigation = self._parse_navigation(raw_index_soup)
+
+        self.metadata = None
+        if self.category_id:
+            topics = self.get_all_topics_category()
+            self.metadata = self._parse_metadata(
+                self._replace_links(raw_index_soup, topics)
+            )
+
+    def _replace_notifications(self, soup):
+        """
+        Given some BeautifulSoup of a document,
+        replace blockquotes with the appropriate notification markup.
+
+        E.g. the following Markdown in a Discourse topic:
+
+            > ⓘ Content
+
+        Will generate the following markup, as per the CommonMark spec
+        (https://spec.commonmark.org/0.29/#block-quotes):
+
+            <blockquote><p>ⓘ Content</p></blockquote>
+
+        Becomes:
+
+            <div class="p-notification">
+                <div class="p-notification__response">
+                    <p class="u-no-padding--top u-no-margin--bottom">
+                        Content
+                    </p>
+                </div>
+            </div>
+        """
+
+        notification_html = (
+            "<div class='{{ notification_class }}'>"
+            "<div class='p-notification__response'>"
+            "{{ contents | safe }}"
+            "</div></div>"
+        )
+
+        notification_template = Template(notification_html)
+        for note_string in soup.findAll(text=re.compile("ⓘ ")):
+            first_paragraph = note_string.parent
+            blockquote = first_paragraph.parent
+            last_paragraph = blockquote.findChildren(recursive=False)[-1]
+
+            if first_paragraph.name == "p" and blockquote.name == "blockquote":
+                # Remove extra padding/margin
+                first_paragraph.attrs["class"] = "u-no-padding--top"
+                if last_paragraph.name == "p":
+                    if "class" in last_paragraph.attrs:
+                        last_paragraph.attrs["class"] += " u-no-margin--bottom"
+                    else:
+                        last_paragraph.attrs["class"] = "u-no-margin--bottom"
+
+                # Remove control emoji
+                notification_html = blockquote.encode_contents().decode(
```

Comment:
this overrides the value of `notification_html` on line 82 of the diff

Reply:
I'm not sure I understand this comment, but see my reply on https://github.com/canonical-web-and-design/maas.io/pull/637/files#r745863373

I didn't modify this code from the original code.

Reply:
ah, you knew that you'd copy/pasted - I didn't :)